<a href="https://colab.research.google.com/github/0IBSIP/0IBSIP/blob/main/Task3_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TASK 3 — DATA CLEANING

## Professional Data Cleaning & Preprocessing of the Titanic Dataset using Python

**Tools:** Python, Pandas, NumPy, Google Colab

### Objective
To transform a raw and potentially inconsistent dataset into a clean,
validated, and analysis-ready dataset by systematically identifying,
handling, and documenting common data-quality issues.

## 1. Introduction

Data cleaning is a critical step in every data analytics and machine
learning workflow. Real-world datasets may contain missing values,
duplicate records, inconsistent formatting, incorrect data types, and
unusual observations.

In this project, the Titanic dataset is used to demonstrate a complete
professional data-cleaning workflow using Python, Pandas, and NumPy.

The project focuses not only on cleaning the data, but also on
documenting the reasoning behind each cleaning decision.

The final objective is to produce a reliable and analysis-ready dataset
that can be used for Exploratory Data Analysis (EDA), visualization,
statistical analysis, feature engineering, and machine learning.

## 2. Project Objectives

The main objectives of this project are:

- Load and inspect the raw dataset.
- Generate a comprehensive data-quality report.
- Identify missing values and determine appropriate treatment strategies.
- Detect and remove duplicate records.
- Standardize inconsistent categorical and text values.
- Correct inappropriate data types.
- Detect numerical outliers using the IQR method.
- Decide whether detected outliers should be retained or removed.
- Perform logical data validation.
- Generate a before-vs-after quality comparison.
- Save the cleaned dataset as a new CSV file.
- Document every major cleaning decision.

## 3. Technology Stack

| Technology | Purpose |
|---|---|
| Python | Programming and data processing |
| Pandas | Data manipulation and cleaning |
| NumPy | Numerical operations |
| Google Colab | Development environment |
| CSV | Dataset input/output |

In [23]:
# ============================================================
# CELL 1 — IMPORT LIBRARIES
# ============================================================

import pandas as pd
import numpy as np
import warnings

warnings.filterwarnings("ignore")

print("Libraries imported successfully!")

Libraries imported successfully!


## 4. Dataset Loading

The raw Titanic dataset is stored in the file `train.csv`.

The dataset is loaded using Pandas. Before performing any cleaning,
the original dataset is preserved so that the effect of each cleaning
step can be measured.

### Raw Dataset File

`train.csv`

In [24]:
# ============================================================
# CELL 2 — LOAD DATASET
# ============================================================

import pandas as pd

# Directly load the uploaded dataset
file_name = "train.csv"

df = pd.read_csv(file_name)

print("Dataset loaded successfully!")
print("File Name:", file_name)
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

display(df.head())

Dataset loaded successfully!
File Name: train.csv
Rows: 891
Columns: 12


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## 5. Dataset Overview

The Titanic dataset contains passenger-level information such as:

- Passenger ID
- Survival status
- Passenger class
- Passenger name
- Sex
- Age
- Number of siblings/spouses
- Number of parents/children
- Ticket
- Fare
- Cabin
- Port of embarkation

The initial dataset contains 891 rows and 12 columns.

In [25]:
# ============================================================
# CELL 3 — BASIC DATASET INFORMATION
# ============================================================

print("========== DATASET OVERVIEW ==========")

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

print("\nColumn Names:")
print(df.columns.tolist())

print("\nData Types:")
display(df.dtypes.to_frame("Data Type"))

print("\nFirst 5 Rows:")
display(df.head())

========== DATASET OVERVIEW ==========
Rows: 891
Columns: 12

Column Names:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']

Data Types:


,Data Type
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64



First 5 Rows:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## 6. Data Quality Assessment

Before cleaning the dataset, a data-quality assessment is performed.

The assessment checks:

- Missing values
- Missing-value percentage
- Data types
- Unique values
- Duplicate records

This establishes a baseline against which the final cleaned dataset
can be compared.

In [26]:
# ============================================================
# CELL 4 — DATA QUALITY REPORT
# ============================================================

quality_report = pd.DataFrame({
    "Column": df.columns,
    "Data_Type": df.dtypes.astype(str).values,
    "Null_Count": df.isnull().sum().values,
    "Null_Percentage": (df.isnull().mean() * 100).round(2).values,
    "Unique_Values": df.nunique().values
})

print("========== DATA QUALITY REPORT ==========")

display(quality_report)

print("\n========== SUMMARY ==========")
print("Total Rows        :", len(df))
print("Total Columns     :", len(df.columns))
print("Total Null Values :", df.isnull().sum().sum())
print("Duplicate Rows    :", df.duplicated().sum())

========== DATA QUALITY REPORT ==========


,Column,Data_Type,Null_Count,Null_Percentage,Unique_Values
0,PassengerId,int64,0,0.00,891
1,Survived,int64,0,0.00,2
2,Pclass,int64,0,0.00,3
3,Name,object,0,0.00,891
4,Sex,object,0,0.00,2
5,Age,float64,177,19.87,88
6,SibSp,int64,0,0.00,7
7,Parch,int64,0,0.00,7
8,Ticket,object,0,0.00,681
9,Fare,float64,0,0.00,248



========== SUMMARY ==========
Total Rows        : 891
Total Columns     : 12
Total Null Values : 866
Duplicate Rows    : 0


## 7. Numerical Range Analysis

Numerical columns are examined to identify unusual or potentially
invalid values.

The analysis includes:

- Minimum
- Maximum
- Mean
- Median
- Standard deviation

This provides an initial understanding of the numerical distribution
before outlier detection.

In [27]:
# ============================================================
# CELL 5 — NUMERIC RANGE REPORT
# ============================================================

numeric_cols = df.select_dtypes(include=np.number).columns

range_report = pd.DataFrame({
    "Column": numeric_cols,
    "Minimum": [df[col].min() for col in numeric_cols],
    "Maximum": [df[col].max() for col in numeric_cols],
    "Mean": [df[col].mean() for col in numeric_cols],
    "Median": [df[col].median() for col in numeric_cols],
    "Std_Dev": [df[col].std() for col in numeric_cols]
})

display(range_report)

,Column,Minimum,Maximum,Mean,Median,Std_Dev
0,PassengerId,1.00,891.0000,446.000000,446.0000,257.353842
1,Survived,0.00,1.0000,0.383838,0.0000,0.486592
2,Pclass,1.00,3.0000,2.308642,3.0000,0.836071
3,Age,0.42,80.0000,29.699118,28.0000,14.526497
4,SibSp,0.00,8.0000,0.523008,0.0000,1.102743
5,Parch,0.00,6.0000,0.381594,0.0000,0.806057
6,Fare,0.00,512.3292,32.204208,14.4542,49.693429


## 8. Establishing the Before-Cleaning Baseline

Before modifying the dataset, important quality metrics are stored.

These metrics will later be compared with the cleaned dataset to
quantify the impact of the cleaning process.

In [28]:
# ============================================================
# CELL 6 — BEFORE-CLEANING METRICS
# ============================================================

before_rows = len(df)
before_columns = len(df.columns)
before_nulls = df.isnull().sum().sum()
before_duplicates = df.duplicated().sum()

print("========== BEFORE CLEANING ==========")
print("Rows              :", before_rows)
print("Columns           :", before_columns)
print("Total Null Values :", before_nulls)
print("Duplicate Rows    :", before_duplicates)

========== BEFORE CLEANING ==========
Rows              : 891
Columns           : 12
Total Null Values : 866
Duplicate Rows    : 0


## 9. Duplicate Detection and Removal

Duplicate records can cause observations to be counted multiple times
and may affect statistical analysis.

Exact duplicate rows are therefore identified and removed.

The number of duplicate rows removed is documented for transparency.

In [29]:
# ============================================================
# CELL 7 — DUPLICATE REMOVAL
# ============================================================

duplicate_count = df.duplicated().sum()

print("Duplicate rows found:", duplicate_count)

if duplicate_count > 0:
    df = df.drop_duplicates().reset_index(drop=True)

print("Duplicate rows removed:", duplicate_count)
print("Rows after duplicate removal:", len(df))

Duplicate rows found: 0
Duplicate rows removed: 0
Rows after duplicate removal: 891


## 10. Column Name Standardization

Consistent column names make the dataset easier to work with and reduce
errors during analysis.

The column names are:

- converted to lowercase
- stripped of unnecessary spaces
- standardized using underscores

In [30]:
# ============================================================
# CELL 8 — STANDARDIZE COLUMN NAMES
# ============================================================

df.columns = (
    df.columns
      .str.strip()
      .str.lower()
      .str.replace(" ", "_")
)

print("Standardized column names:")
print(df.columns.tolist())

Standardized column names:
['passengerid', 'survived', 'pclass', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket', 'fare', 'cabin', 'embarked']


## 11. Text and Categorical Standardization

Inconsistent text formatting can result in duplicate categories.

For example:

`Male`, `male`, and `M`

may represent the same category.

Text fields are therefore stripped of unnecessary spaces and important
categorical values are standardized.

In [31]:
# ============================================================
# CELL 9 — STANDARDIZE TEXT VALUES
# ============================================================

text_columns = [
    "name",
    "sex",
    "ticket",
    "cabin",
    "embarked"
]

for col in text_columns:
    df[col] = df[col].astype("string").str.strip()

# Standardize Sex
df["sex"] = (
    df["sex"]
    .str.lower()
    .replace({
        "m": "male",
        "male": "male",
        "f": "female",
        "female": "female"
    })
)

# Standardize Embarked
df["embarked"] = df["embarked"].str.upper()

print("Text standardization completed.")

Text standardization completed.


## 12. Missing Value Analysis

Missing values are inspected before treatment.

Different columns require different strategies depending on their
data type and meaning.

The following strategies are used:

| Column | Strategy | Reason |
|---|---|---|
| Age | Median | Numerical variable and robust to extreme values |
| Embarked | Mode | Categorical variable |
| Cabin | Unknown | Preserve observations and retain missingness information |

In [32]:
# ============================================================
# CELL 10 — MISSING VALUE REPORT
# ============================================================

missing_report = pd.DataFrame({
    "Column": df.columns,
    "Missing_Values": df.isnull().sum().values,
    "Missing_Percentage":
        (df.isnull().mean() * 100).round(2).values
})

missing_report = missing_report[
    missing_report["Missing_Values"] > 0
]

print("========== MISSING VALUES BEFORE TREATMENT ==========")

display(missing_report)

========== MISSING VALUES BEFORE TREATMENT ==========


,Column,Missing_Values,Missing_Percentage
5,age,177,19.87
10,cabin,687,77.10
11,embarked,2,0.22


In [33]:
# ============================================================
# CELL 11 — MISSING VALUE TREATMENT
# ============================================================

# Age → Median
age_median = df["age"].median()

df["age"] = df["age"].fillna(age_median)

# Embarked → Mode
embarked_mode = df["embarked"].mode()[0]

df["embarked"] = df["embarked"].fillna(embarked_mode)

# Cabin → Unknown
df["cabin"] = df["cabin"].fillna("Unknown")

print("Missing value treatment completed.")

print("\nRemaining missing values:")
display(df.isnull().sum())

Missing value treatment completed.

Remaining missing values:


,0
passengerid,0
survived,0
pclass,0
name,0
sex,0
age,0
sibsp,0
parch,0
ticket,0
fare,0


## 13. Data Type Correction

Correct data types are important for reliable analysis.

`PassengerId` is an identifier, so it is treated as a string rather
than a numerical measurement.

Numerical variables are converted to appropriate numerical types and
text variables are stored as strings.

In [34]:
# ============================================================
# CELL 12 — DATA TYPE CORRECTION
# ============================================================

# Identifier
df["passengerid"] = df["passengerid"].astype("string")

# Text columns
for col in ["name", "sex", "ticket", "cabin", "embarked"]:
    df[col] = df[col].astype("string")

# Numeric columns
numeric_columns = [
    "survived",
    "pclass",
    "age",
    "sibsp",
    "parch",
    "fare"
]

for col in numeric_columns:
    df[col] = pd.to_numeric(
        df[col],
        errors="coerce"
    )

print("Data types corrected successfully.")

display(df.dtypes.to_frame("Final Data Type"))

Data types corrected successfully.


,Final Data Type
passengerid,string[python]
survived,int64
pclass,int64
name,string[python]
sex,string[python]
age,float64
sibsp,int64
parch,int64
ticket,string[python]
fare,float64


## 14. Outlier Detection Using IQR

The Interquartile Range (IQR) method is used to identify statistical
outliers in numerical columns.

The analysis is performed on:

- Age
- SibSp
- Parch
- Fare

For each variable, Q1, Q3, IQR, lower bound, upper bound, and number
of detected outliers are calculated.

In [35]:
# ============================================================
# CELL 13 — IQR OUTLIER DETECTION
# ============================================================

outlier_columns = [
    "age",
    "sibsp",
    "parch",
    "fare"
]

outlier_results = []

for col in outlier_columns:

    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)

    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outlier_mask = (
        (df[col] < lower_bound) |
        (df[col] > upper_bound)
    )

    outlier_count = outlier_mask.sum()

    outlier_results.append({
        "Column": col,
        "Q1": round(Q1, 2),
        "Q3": round(Q3, 2),
        "IQR": round(IQR, 2),
        "Lower_Bound": round(lower_bound, 2),
        "Upper_Bound": round(upper_bound, 2),
        "Outlier_Count": int(outlier_count),
        "Outlier_Percentage":
            round(outlier_count / len(df) * 100, 2)
    })

outlier_report = pd.DataFrame(outlier_results)

print("========== IQR OUTLIER REPORT ==========")

display(outlier_report)

========== IQR OUTLIER REPORT ==========


,Column,Q1,Q3,IQR,Lower_Bound,Upper_Bound,Outlier_Count,Outlier_Percentage
0,age,22.00,35.0,13.00,2.50,54.50,66,7.41
1,sibsp,0.00,1.0,1.00,-1.50,2.50,46,5.16
2,parch,0.00,0.0,0.00,0.00,0.00,213,23.91
3,fare,7.91,31.0,23.09,-26.72,65.63,116,13.02


## 15. Outlier Treatment Decision

The IQR method identifies statistical outliers, but an outlier is not
automatically an incorrect value.

For this dataset, detected outliers are retained because extreme values
may represent legitimate passenger characteristics.

For example, a high Fare value may represent a genuine higher-priced
ticket.

Therefore, the project does not delete statistical outliers without
evidence that they are invalid.

This approach reduces unnecessary information loss.

## 16. Logical Data Validation

After cleaning, logical validation checks are performed.

The purpose is to identify values that are impossible or inconsistent
with the expected meaning of the variables.

In [36]:
# ============================================================
# CELL 14 — LOGICAL VALIDATION
# ============================================================

print("========== LOGICAL VALIDATION ==========")

print(
    "Negative Age values:",
    (df["age"] < 0).sum()
)

print(
    "Negative Fare values:",
    (df["fare"] < 0).sum()
)

print(
    "Invalid Pclass values:",
    (~df["pclass"].isin([1, 2, 3])).sum()
)

print(
    "Invalid Survived values:",
    (~df["survived"].isin([0, 1])).sum()
)

print(
    "Invalid Embarked values:",
    (~df["embarked"].isin(["C", "Q", "S"])).sum()
)

========== LOGICAL VALIDATION ==========
Negative Age values: 0
Negative Fare values: 0
Invalid Pclass values: 0
Invalid Survived values: 0
Invalid Embarked values: 0


## 17. Final Missing Value Validation

After applying the selected missing-value strategies, the dataset is
checked again.

The goal is to verify that the intended missing values have been
successfully handled.

In [37]:
# ============================================================
# CELL 15 — FINAL NULL CHECK
# ============================================================

final_null_report = pd.DataFrame({
    "Column": df.columns,
    "Null_Count": df.isnull().sum().values,
    "Null_Percentage":
        (df.isnull().mean() * 100).round(2).values
})

print("========== FINAL NULL REPORT ==========")

display(final_null_report)

print(
    "Total Remaining Null Values:",
    df.isnull().sum().sum()
)

========== FINAL NULL REPORT ==========


,Column,Null_Count,Null_Percentage
0,passengerid,0,0.0
1,survived,0,0.0
2,pclass,0,0.0
3,name,0,0.0
4,sex,0,0.0
5,age,0,0.0
6,sibsp,0,0.0
7,parch,0,0.0
8,ticket,0,0.0
9,fare,0,0.0


Total Remaining Null Values: 0


In [38]:
# ============================================================
# CELL 16 — FINAL DUPLICATE CHECK
# ============================================================

final_duplicates = df.duplicated().sum()

print("========== FINAL DUPLICATE CHECK ==========")
print("Remaining Duplicate Rows:", final_duplicates)

========== FINAL DUPLICATE CHECK ==========
Remaining Duplicate Rows: 0


## 18. Before vs After Cleaning

A before-vs-after comparison is created to measure the effect of the
data-cleaning process.

The comparison focuses on:

- Row count
- Column count
- Total null values
- Duplicate rows

This provides measurable evidence of the cleaning results.

In [39]:
# ============================================================
# CELL 17 — BEFORE VS AFTER SUMMARY
# ============================================================

after_rows = len(df)
after_columns = len(df.columns)
after_nulls = df.isnull().sum().sum()
after_duplicates = df.duplicated().sum()

summary = pd.DataFrame({
    "Metric": [
        "Row Count",
        "Column Count",
        "Total Null Values",
        "Duplicate Rows"
    ],

    "Before Cleaning": [
        before_rows,
        before_columns,
        before_nulls,
        before_duplicates
    ],

    "After Cleaning": [
        after_rows,
        after_columns,
        after_nulls,
        after_duplicates
    ]
})

print("========== BEFORE VS AFTER SUMMARY ==========")

display(summary)

========== BEFORE VS AFTER SUMMARY ==========


,Metric,Before Cleaning,After Cleaning
0,Row Count,891,891
1,Column Count,12,12
2,Total Null Values,866,0
3,Duplicate Rows,0,0


In [40]:
# ============================================================
# CELL 18 — FINAL DATA QUALITY REPORT
# ============================================================

final_quality_report = pd.DataFrame({
    "Column": df.columns,
    "Final_Data_Type": df.dtypes.astype(str).values,
    "Null_Count": df.isnull().sum().values,
    "Unique_Values": df.nunique().values
})

print("========== FINAL DATA QUALITY REPORT ==========")

display(final_quality_report)

========== FINAL DATA QUALITY REPORT ==========


,Column,Final_Data_Type,Null_Count,Unique_Values
0,passengerid,string,0,891
1,survived,int64,0,2
2,pclass,int64,0,3
3,name,string,0,891
4,sex,string,0,2
5,age,float64,0,88
6,sibsp,int64,0,7
7,parch,int64,0,7
8,ticket,string,0,681
9,fare,float64,0,248


## 19. Cleaned Dataset Preview

The cleaned dataset is displayed to verify that the transformations
have been successfully applied.

In [41]:
# ============================================================
# CELL 19 — CLEANED DATASET PREVIEW
# ============================================================

print("========== CLEANED DATASET ==========")

print("Final Shape:", df.shape)

display(df.head(10))

========== CLEANED DATASET ==========
Final Shape: (891, 12)


,passengerid,survived,pclass,name,sex,age,sibsp,parch,ticket,fare,cabin,embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,Unknown,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,Unknown,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,Unknown,S
5,6,0,3,"Moran, Mr. James",male,28.0,0,0,330877,8.4583,Unknown,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,Unknown,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,Unknown,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,Unknown,C


## 20. Export Cleaned Dataset

The cleaned dataset is saved as a new CSV file.

The original `train.csv` file is not overwritten. This preserves the
raw dataset and makes the cleaning workflow reproducible.

In [42]:
# ============================================================
# CELL 20 — SAVE CLEANED DATASET
# ============================================================

output_file = "cleaned_dataset_task3.csv"

df.to_csv(
    output_file,
    index=False
)

print("Cleaned dataset saved successfully!")
print("Output file:", output_file)

Cleaned dataset saved successfully!
Output file: cleaned_dataset_task3.csv


In [43]:
# ============================================================
# CELL 21 — DOWNLOAD CLEANED DATASET
# ============================================================

from google.colab import files

files.download("cleaned_dataset_task3.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## 21. Recruiter-Focused Project Highlights

### Technical Highlights

- Developed an end-to-end data-cleaning pipeline using Python,
  Pandas, and NumPy.
- Performed systematic data-quality assessment before transformation.
- Applied column-specific missing-value treatment.
- Used median and mode imputation appropriately.
- Preserved missing Cabin information using an `Unknown` category.
- Detected and removed exact duplicate records.
- Standardized categorical and text values.
- Corrected identifier and numerical data types.
- Implemented IQR-based outlier detection.
- Performed logical data validation.
- Created a measurable before-vs-after quality report.
- Exported a separate analysis-ready CSV dataset.

### Analytical Thinking

The project demonstrates the ability to move from:

**Problem → Investigation → Cleaning Decision → Validation → Final Dataset**

The cleaning process focuses not only on writing code, but also on
understanding the meaning of the data and justifying each transformation.

# 22. Conclusion

This project successfully demonstrates a professional and reproducible
data-cleaning workflow using the Titanic dataset.

The raw dataset was systematically inspected to identify missing values,
duplicate records, inconsistent categorical values, data-type issues,
potential outliers, and logical anomalies.

Appropriate strategies were then applied based on the characteristics
of individual variables. Numerical missing values in `Age` were handled
using median imputation, categorical missing values in `Embarked` were
handled using mode imputation, and missing `Cabin` information was
represented using an `Unknown` category.

Duplicate records were checked and removed where necessary. Text and
categorical values were standardized, while identifiers and numerical
variables were assigned appropriate data types.

Potential numerical outliers were detected using the IQR method.
Rather than automatically deleting extreme observations, they were
retained because they may represent legitimate passenger information.

Finally, logical validation, final quality checks, and a before-vs-after
comparison were performed. The cleaned dataset was exported as:

`cleaned_dataset_task3.csv`

The resulting dataset is now suitable for downstream Exploratory Data
Analysis (EDA), visualization, statistical analysis, feature
engineering, and machine-learning workflows.

### Final Learning Outcome

This project demonstrates that professional data cleaning is not simply
about removing missing or unusual values. It involves:

**Understanding the data → Identifying quality issues → Choosing
appropriate treatments → Validating the results → Documenting decisions
→ Producing a reliable analysis-ready dataset.**